### PR TITLE
Customer Management Admin Client

### LINKED ISSUE #9: Company customer management
Client should allow CRUD operations for a companies client base

### PR DESCRIPTION
**Pull request by DominikWas**
fixes #9, #10 (not #11!)
10, 11 = children of 9

### CODE REVIEW

File: AdminClient/src/app/customers.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+describe('CustomersService', () => {
+  let service: CustomersService;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({});
+    service = TestBed.inject(CustomersService);
+  });
+
+  it('should be created', () => {
```
Comment (markscamilleri) [inline AdminClient/src/app/customers.service.spec.ts:13]: This test alone is not enough. Please test the public methods

File: AdminClient/src/app/customers/dialog-body/dialog-body.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(DialogBodyComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
```
Comment (markscamilleri) [inline AdminClient/src/app/customers/dialog-body/dialog-body.component.spec.ts:22]: Test the public methods

File: AdminClient/src/app/mock-customers.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { Customer} from './customer';
+
+export const CUSTOMERS: Customer[] = [
+  { id: 1, name: 'Bang Pharmaceuticals', category: 'Business', telephone: '02079460674', email: 'lmywelle@bangpharma.com',
```
Comment (markscamilleri) [inline AdminClient/src/app/mock-customers.ts:4]: Ngl you had me there for a second I thought this was real data 😆

File: AdminClient/src/app/customers/customers.component.html [hunk 4]
```diff
@@ -0,0 +1,82 @@
+<div class="container"
+     fxLayout
+     fxLayout.xs="column"
+     fxLayoutAlign="center stretch"
+     fxLayoutGap="10px"
+     fxLayoutGap.xs="0">
+
+  <div id="table">
```
Comment (HarryPi) [inline AdminClient/src/app/customers/customers.component.html:8]: Please review the style for this i know you said this is a skeleton but you should be using mat-table module since you are importing it as well.
See https://material.angular.io/components/table/overview
It will save you a lot of trouble trying to find the right CSS as it comes with some nice out of the box design. currently putting it politely this hurts my eyes.

File: AdminClient/src/app/customers/customers.component.html [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        <td>{{ customer.country }}</td>
+        <td>{{ customer.postcode }}</td>
+      </tr>
+      </tbody>
+    </table>
+  </div>
+
+</div>
+
+<div class="container"
```
Comment (HarryPi) [inline AdminClient/src/app/customers/customers.component.html:47]: If possible transfer this to a toolbar format rather than just throwing them out there. Again see what the Angular team offers you out of the box https://material.angular.io/components/toolbar/overview so it wont take you time to set it up. Also see how you will integrate this items together i.e Will this be a pop-up? Will this cover the entire screen?

File: AdminClient/src/app/customers/customers.component.html (hunk 4 above)
Comment (HarryPi) [reply AdminClient/src/app/customers/customers.component.html:8]: Another comment here separate to the above. Ive noticed your table just "appears" as a user experience this is not good. there is no indication that something is going on in the background, thus if your table takes longer than expected the user will consider this broken or just spam the refresh button. Please consider a loading indication.

File: AdminClient/src/app/customers/dialog-body/dialog-body.component.html [hunk 6]
```diff
@@ -0,0 +1,144 @@
+<div class="container" fxLayout
+     fxLayout.xs="column"
+     fxLayoutAlign="none"
+     fxLayoutGap="10px"
+     fxLayoutGap.xs="0">
+
+  <div id="create" fxFlex fxFlexOffset="50px" *ngIf="displayCreate()">
```
Comment (HarryPi) [inline AdminClient/src/app/customers/dialog-body/dialog-body.component.html:7]: Again see https://material.angular.io/components/input/overview to quickly give your inputs some nice styling that wont take long and can be considered an MVP. In my opinion having something presentable for an MVP is important as it otherwise distract the user from seeing the functionality @markscamilleri please share your input in this in case styling is not important for MVP.

File: AdminClient/src/app/customers/dialog-body/dialog-body.component.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+@Component({
+  selector: 'app-dialog-body',
+  templateUrl: './dialog-body.component.html',
+  styleUrls: ['./dialog-body.component.scss']
+})
+export class DialogBodyComponent implements OnInit {
+
+  dialogForm = new FormGroup({
+    dialogControl1: new FormControl(''),
```
Comment (HarryPi) [inline AdminClient/src/app/customers/dialog-body/dialog-body.component.ts:16]: Please give this some sort of meaningful naming what differences is dialogControl1 from dialogControl10? How will you know when debugging?

File: AdminClient/src/app/customers/dialog-body/dialog-body.component.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+          .subscribe();
+      }
+
+      if (this.data === 'delete') { // update customer
+        this.customerService.deleteCustomerRecord(this.dialogForm.get('dialogControl1').value.toString())
+          .subscribe();
+      }
+    }
+  }
+}
```
Comment (HarryPi) [inline AdminClient/src/app/customers/dialog-body/dialog-body.component.ts:100]: Where is the validation for the forms?

File: AdminClient/src/app/customers/dialog-body/dialog-body.component.spec.ts (hunk 2 above)
Comment (DominikWas) [reply AdminClient/src/app/customers/dialog-body/dialog-body.component.spec.ts:22]: Will add the tests on, sorry forgot to add test cases.

File: AdminClient/src/app/customers/dialog-body/dialog-body.component.ts (hunk 8 above)
Comment (DominikWas) [reply AdminClient/src/app/customers/dialog-body/dialog-body.component.ts:100]: I was not able to add them on time, will be up on next commit though.

File: AdminClient/src/app/customers/customers.component.html (hunk 4 above)
Comment (DominikWas) [reply AdminClient/src/app/customers/customers.component.html:8]: I will add improved styling, will also take a look at loading options.

File: AdminClient/src/app/customers/dialog-body/dialog-body.component.ts (hunk 7 above)
Comment (HarryPi) [reply AdminClient/src/app/customers/dialog-body/dialog-body.component.ts:16]: It doesnt matter if its reused. It should be reused for forms that make sense to be re-used together otherwise your validation will be all messed up. For example new user and edit user can be grouped together yes; But then New user and new user role cant.
So example new user / edit user can have fields such as
```
 email: new FormControl('', [Validators.email, Validators.required]),
 password: new FormControl('', [Validators.required, Validators.minLength(7)]),
 address: new FormControl('', [Validators.required]),
 name: new FormControl('', [Validators.pattern([a-zA-Z ]*), Validators.required])
```
on the other hand say role form should look something like
```
 email: new FormControl('', [Validators.email, Validators.required]),
 role: new FormControl('', [Validators.validateRole), Validators.required)
```
As you can see the validations dont match.
Also assume you are debugging how will you know which field was assigned to controller1-12 etc on the other hand if it had a meaningfull name like email, role, password that would be easy to understand. 
**Re-using code is good and you should aim for it but only where it makes sense**

File: AdminClient/src/app/customers/customers.component.html (hunk 4 above)
Comment (HarryPi) [reply AdminClient/src/app/customers/customers.component.html:8]: Also keep in mind that the Angular material components offer a lot of extra functionality. Which means you wont have to implement it.

File: AdminClient/src/app/customer.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+export interface Customer {
```
Comment (markscamilleri) [inline AdminClient/src/app/customer.ts:1]: Should this be in it's own folder? (like `model`) @HarryPi?

Comment (markscamilleri) [inline AdminClient/src/app/customer.ts:1]: Should this be in it's own folder? (like `model`)?

File: AdminClient/src/app/customers.service.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
+import { Injectable } from '@angular/core';
```
Comment (markscamilleri) [inline AdminClient/src/app/customers.service.ts:1]: Can you move this and the tests to a `services/customers` path? (`AdminClient/src/app/services/customers/customers.service.ts`)

File: AdminClient/src/app/customers/dialog-body/dialog-body.component.ts (hunk 7 above)
Comment (markscamilleri) [reply AdminClient/src/app/customers/dialog-body/dialog-body.component.ts:16]: In that case, the file and class should also be renamed from `dialog-body`.

File: AdminClient/src/styles.scss [hunk 11]
```diff
@@ -1,32 +1,16 @@
-
```
Comment (markscamilleri) [inline AdminClient/src/styles.scss:1]: Merge master into this cause this was updated.

File: AdminClient/src/app/customers/customers.component.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
+import {Component, Inject, OnInit, Output} from '@angular/core';
```
Comment (markscamilleri) [inline AdminClient/src/app/customers/customers.component.ts:1]: @HarryPi should this be in its own module?

Comment (HarryPi) [reply AdminClient/src/app/customers/customers.component.ts:1]: A component should be its own module when its a top level component i.e not a child of any component besides the app component. Giving the definition as i dont know the structure here.

File: AdminClient/src/app/customers/customers.component.spec.ts [hunk 13]
```diff
@@ -22,4 +22,45 @@ describe('CustomersComponent', () => {
...
+    const inputData = 'load';
+    component.data = inputData;
+    expect(component.data).toBe('load');
+  });
+
+  it('should call the update record dialog once', () => {
+    spyOn(component, 'openUpdateRecordDialog')
+    component.openUpdateRecordDialog()
+    expect(component.openUpdateRecordDialog).toHaveBeenCalledTimes(1)
+  });
```
Comment (HarryPi) [inline AdminClient/src/app/customers/customers.component.spec.ts:65]:
```suggestion

```
This was a good attempt in introducing unit tests however some things to take into consideration.

For example
```
  component.customers = [{id: 1, name: 'a', category: 'a', telephone: 'a', email: 'a',
      website: 'a', contact: 'a', address: 'a', townOrCity: 'a',
      county: 'a', country: 'a', postcode: 'a'}];
    component.customers.pop();
```

What you are testing here is that JavaScript is working that is expected. You simply add an element to the array then use the **array** function to remove the element.

What we want to do when testing is change the state whether that is in our app or a more general state i.e NgRx and then check if for example the HTML has changed to reflect that change. Or something else we could be testing is that a custom function we made returns the expected result or triggers the subsequent actions that must follow.

Tl;dr what you are doing here is testing that JavaScript works.

Comment (DominikWas) [reply AdminClient/src/app/customers/customers.component.spec.ts:65]: Yes sorry, I was waiting for your table changes so I can check the resulting HTML. Will be added

File: AdminClient/src/app/customers.service.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    console.log(CUSTOMERS[CUSTOMERS.length - 1])
+    return of(CUSTOMERS[CUSTOMERS.length - 1]);
+  }
+
+  updateCustomerRecord(idOrName: any, column: string, newEntry: string): Observable<any> {
+    let i = 0;
+    if (idOrName.endsWith('*')) { // if *, search by id
+      while (i < CUSTOMERS.length) { // scan customer records
+        if (idOrName === CUSTOMERS[i].id && CUSTOMERS[i].hasOwnProperty(column)) { // check if the id exists + column is correct
+          console.log('Property: ' + column + ' exists!')
```
[the remark this answers is not in the document]
Comment (DominikWas) [reply AdminClient/src/app/customers.service.ts]: Yes logging was just for debugging, removed now.

File: AdminClient/src/app/customers/customers.component.scss [hunk 15]
```diff
@@ -0,0 +1,13 @@
+md-button {
```
[the remark this answers is not in the document]
Comment (DominikWas) [reply AdminClient/src/app/customers/customers.component.scss:1]: Good idea, will be added

**Comment by DominikWas** [comment]
To Do:
* Loader/Prog. Spinner
* Actions
* Tests (If possible - low priority at the moment)

File: .stylelintignore [hunk 16]
```diff
@@ -1 +1,2 @@
 RepClient/src/global.scss
+AdminClient/src/styles.scss
```
Comment (markscamilleri) [inline .stylelintignore:2]: Should we ignore whole global files?

**Comment by markscamilleri** [comment]
This broke tests (https://travis-ci.com/github/Quantium-Nonsense/ProjectOmega/jobs/322337755)

File: .stylelintignore (hunk 16 above)
Comment (HarryPi) [reply .stylelintignore:2]: Yeah coz angular / codebot wont pick auto generated classes like in for loops etc and then we would have to allow for some really stupid rules. So i tend to create those only in there and ignore it all together but its up for discussion

**Comment by HarryPi** [comment]
> This broke tests (https://travis-ci.com/github/Quantium-Nonsense/ProjectOmega/jobs/322337755)

Yh fix was incoming just had a minor delivery inconvenience
